pkg/reconciler/common/prune.go

```diff
@@ -0,0 +1,162 @@
+		}
+		pru := tC.Spec.Pruner
+		logger := logging.FromContext(ctx)
+		ownerRef := v1.OwnerReference{
+			APIVersion: ownerAPIVer,
+			Kind:       ownerKind,
+			Name:       tC.Name,
+			UID:        tC.ObjectMeta.UID,
+		}
+
+		pruningNamespaces, err := GetPrunableNamespaces(k, ctx)
+		if err != nil {
+			return err
+		}
+		if err := createCronJob(k, ctx, pru, tC.Spec.TargetNamespace, pruningNamespaces, ownerRef, tknImage); err != nil {
+			logger.Error("failed to create cronjob ", err)
+
+		}
+	}
+	return nil
+}
+
+func GetPrunableNamespaces(k kubernetes.Interface, ctx context.Context) ([]string, error) {
+	nsList, err := k.CoreV1().Namespaces().List(ctx, v1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	var allNameSpaces []string
+	for _, ns := range nsList.Items {
+		if !strings.HasPrefix(ns.Name, openshiftPrefix) && !strings.HasPrefix(ns.Name, kubePrefix) {
+			allNameSpaces = append(allNameSpaces, ns.Name)
+		}
+	}
+	return allNameSpaces, nil
+}
+
+func createCronJob(k kubernetes.Interface, ctx context.Context, pru v1alpha1.Prune, targetNs string, pruningNs []string, oR v1.OwnerReference, tknImage string) error {
+	pruneContainers := getPruningContainers(pru.Resources, pruningNs, pru.Keep, tknImage)
+	backOffLimit := int32(3)
+	ttlSecondsAfterFinished := int32(3600)
+	cj := &v1beta1.CronJob{
+		TypeMeta: v1.TypeMeta{
+			Kind:       "CronJob",
+			APIVersion: "batch/v1beta1",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name:            CronName,
+			OwnerReferences: []v1.OwnerReference{oR},
+		},
+		Spec: v1beta1.CronJobSpec{
+			Schedule:          pru.Schedule,
+			ConcurrencyPolicy: "Forbid",
+			JobTemplate: v1beta1.JobTemplateSpec{
+
+				Spec: batchv1.JobSpec{
+					TTLSecondsAfterFinished: &ttlSecondsAfterFinished,
+					BackoffLimit:            &backOffLimit,
+
+					Template: corev1.PodTemplateSpec{
+						Spec: corev1.PodSpec{
+							Containers:         pruneContainers,
+							RestartPolicy:      "OnFailure",
+							ServiceAccountName: tektonSA,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	if _, err := k.BatchV1beta1().CronJobs(targetNs).Create(ctx, cj, v1.CreateOptions{}); err != nil {
+		if strings.Contains(err.Error(), "already exists") {
+			if _, err := k.BatchV1beta1().CronJobs(targetNs).Update(ctx, cj, v1.UpdateOptions{}); err != nil {
+				return err
+			}
+		}
+		return err
+	}
+	return nil
+}
+
+func getPruningContainers(resources, namespaces []string, keep int, tknImage string) []corev1.Container {
+	containers := []corev1.Container{}
+
+	for _, res := range resources {
+		for _, ns := range namespaces {
+			cmdArgs := deleteCommand(res, keep, ns)
+			cName := res + "-pruner-tkn-" + ns
+			container := corev1.Container{
+				Name:                     cName,
+				Image:                    tknImage,
+				Command:                  []string{"tkn"},
+				Args:                     cmdArgs,
+				TerminationMessagePolicy: "FallbackToLogsOnError",
+			}
+			containers = append(containers, container)
+		}
+	}
```

Comment:
let us avoid the outer loop by combining the 2 commands.
eg: 
`tkn pr delele --keep 100; tkn tr delete --keep 100`